src/libraries/System.Text.Json/src/System/Text/Json/Serialization/WriteStack.cs

```diff
@@ -125,36 +124,33 @@ public void Push()
                     JsonTypeInfo jsonTypeInfo = Current.GetPolymorphicJsonPropertyInfo().RuntimeTypeInfo;
                     JsonNumberHandling? numberHandling = Current.NumberHandling;
 
-                    AddCurrent();
-                    Current.Reset();
+                    EnsurePushCapacity();
+                    _stack[_count - 1] = Current;
+                    Current = default;
+                    _count++;
 
                     Current.JsonTypeInfo = jsonTypeInfo;
                     Current.DeclaredJsonPropertyInfo = jsonTypeInfo.PropertyInfoForTypeInfo;
                     // Allow number handling on property to win over handling on type.
                     Current.NumberHandling = numberHandling ?? Current.DeclaredJsonPropertyInfo.NumberHandling;
                 }
             }
-            else if (_continuationCount == 1)
-            {
-                // No need for a push since there is only one stack frame.
-                Debug.Assert(_count == 1);
-                _continuationCount = 0;
-            }
             else
             {
-                // A continuation, adjust the index.
-                Current = _previous[_count - 1];
-
-                // Check if we are done.
-                if (_count == _continuationCount)
+                // We are re-entering a continuation, adjust indices accordingly
+                if (_count++ > 0)
                 {
-                    _continuationCount = 0;
+                    Current = _stack[_count - 1];
                 }
-                else
+
+                // check if we are done
+                if (_continuationCount == _count)
                 {
-                    _count++;
+                    _continuationCount = 0;
                 }
             }
+
+            Debug.Assert(PropertyPath() is not null);
```

Comment:
Why was this added? PropertyPath would never be null, even if no path (it would be ".")

Reply:
It's not particularly meaningful, I ony added it to force execution of `PropertyPath()` on every run. My changes for IAsyncEnumerable actually broke this method but we had no test coverage for it. Perhaps it might be clearer if I change it to
```csharp
#if DEBUG
_ = PropertyPath();
#endif
```